p2p/peer.go

```diff
@@ -578,38 +642,35 @@ func (m *PeerManager) DialFailed(peerID NodeID, address PeerAddress) error {
 		}
 	}
 
-	peer, err := m.store.Get(peerID)
-	if err != nil || peer == nil { // Peer may have been removed while dialing, ignore.
-		return err
+	peer, ok := m.store.Get(peerID)
+	if !ok { // Peer may have been removed while dialing, ignore.
+		return nil
 	}
-	addressInfo := peer.LookupAddressInfo(address)
-	if addressInfo == nil {
+	addressInfo, ok := peer.AddressInfo[address.String()]
+	if !ok {
 		return nil // Assume the address has been removed, ignore.
 	}
 	addressInfo.LastDialFailure = time.Now().UTC()
 	addressInfo.DialFailures++
-	if err = m.store.Set(peer); err != nil {
+	if err := m.store.Set(peer); err != nil {
 		return err
 	}
 
 	// We spawn a goroutine that notifies DialNext() again when the retry
 	// timeout has elapsed, so that we can consider dialing it again.
-	//
-	// FIXME: We need to calculate the retry delay outside of the goroutine,
-	// since the arguments are currently pointers to structs shared in the
-	// peerStore. The peerStore should probably return struct copies instead,
-	// to avoid these sorts of issues.
-	if retryDelay := m.retryDelay(peer, addressInfo.DialFailures); retryDelay != time.Duration(math.MaxInt64) {
-		go func() {
-			timer := time.NewTimer(retryDelay)
-			defer timer.Stop()
-			select {
-			case <-timer.C:
-				m.wakeDial()
-			case <-m.closeCh:
-			}
-		}()
-	}
+	go func() {
+		retryDelay := m.retryDelay(addressInfo.DialFailures, peer.Persistent)
+		if retryDelay == time.Duration(math.MaxInt64) {
+			return
+		}
+		timer := time.NewTimer(retryDelay)
+		defer timer.Stop()
+		select {
+		case <-timer.C:
```

Comment:
`time.After`

Reply:
See comment on other PR, that would leak goroutines.